### PR TITLE
Fix question creation for choice questions

### LINKED ISSUE #177: Selected Answer Set & Possible Solution to a Question not saved correctly
- [x] question answering (#178)
- [ ] question creation (#179)

We switched from `:answer/answer` to `:answer/selected-solutions` for single-/multiple-choice questions and similarly for question creation.

https://github.com/pkoerner/expert-parakeet/blob/115ae8091cf96f687a65cc6b7038f9197b751ef9/src/clj/db.clj#L383-L395

https://github.com/pkoerner/expert-parakeet/blob/115ae8091cf96f687a65cc6b7038f9197b751ef9/test/controllers/question/question_controller_test.clj#L125-L197

### PR DESCRIPTION
**Pull request by iTitus**
Closes #177

### CODE REVIEW

**Comment by iTitus** [comment]
Prev: image

New: image

It still looks a bit cramped and the legend/labels are of different sizes, but the functionality is strictly better.

File: src/clj/services/question_service/question_service.clj [hunk 1]
```diff
@@ -71,11 +73,12 @@ an error-set with a specified error is returned. "
   (db/get-all-question-categories (.db this)))
 
 
-(def ^:private question-keys
-  [:question/statement :question/max-points :question/type
-   :question/possible-solutions :question/correct-solutions
-   :question/evaluation-criteria
-   :question/categories])
+(defn- as-coll
```
Comment (davrad) [inline src/clj/services/question_service/question_service.clj:76]: Documentation for function is missing

File: src/clj/services/question_service/question_service.clj [hunk 2]
```diff
@@ -84,160 +87,100 @@ an error-set with a specified error is returned. "
...
-                                 :question-type-as-string string?)
-                     :possible-solutions (s/or :nil nil? :single string? :multiple (s/coll-of string?))
-                     :correct-solutions (s/or :nil nil? :solution string? :multiple-solutions (s/coll-of string?))
-                     :evaluation-criteria (s/or :nil nil? :criteria string?)
-                     :categories (s/or :nil nil? :single string? :multiple (s/coll-of string?)))
-        :ret (s/or :errors #(contains? % :errors)
-                   :question :question/question))
+;; we are using a vector to preserve the iteration order!
+(def ^:private question-validators
+  [[:type (fn [_ value]
```
Comment (davrad) [inline src/clj/services/question_service/question_service.clj:92]: Documentation for function is missing

File: src/clj/services/question_service/question_service.clj [hunk 3]
```diff
@@ -84,160 +87,100 @@ an error-set with a specified error is returned. "
...
-
-  (let [question (parse-question question-statement achivable-points type
-                                 possible-solutions correct-solutions
-                                 evaluation-criteria
-                                 categories)
-        errors (question :errors)]
-    (if errors
-      question
-      (merge question (validate-parsed-question question)))))
+  [_ question-form-data]
```
Comment (davrad) [inline src/clj/services/question_service/question_service.clj:172]: Comment is missing

File: src/clj/views/question/create_question_view.clj [hunk 4]
```diff
@@ -10,196 +10,200 @@
     [util.hiccup-extensions :refer [optional-error-display script]]))
 
 
-(def create-question-error-keys
-  "Possible keys for which errors can be displayed in the `question-form`."
-  #{:question/statement :question/type :question/max-points
-    :question/possible-solutions :question/correct-solutions
-    :question/evaluation-criteria})
+(defn- as-coll
```
Comment (davrad) [inline src/clj/views/question/create_question_view.clj:13]: Same as above

File: src/clj/views/question/create_question_view.clj [hunk 5]
```diff
@@ -10,196 +10,200 @@
...
+  [coll-or-single]
+  (cond
+    (nil? coll-or-single) []
+    (coll? coll-or-single) coll-or-single
+    :else [coll-or-single]))
 
 
 (defn- free-text-inputs
```
Comment (davrad) [inline src/clj/views/question/create_question_view.clj:21]: Comment is missing

File: src/clj/views/question/create_question_view.clj [hunk 6]
```diff
@@ -10,196 +10,200 @@
...
-       solution-error
-       [:label {:for solution-list-id} "Correct answer(/en):"] [:br]
-       [:ul {:id solution-list-id}]])))
+    [:fieldset#free-text
+     [:legend {:class "form-label"} "Evaluation criteria"]
+     (optional-error-display :evaluation-criteria errors)
+     (hform/text-area {:class "form-control"} "evaluation-criteria" (get question-data :evaluation-criteria))]))
+
+
+(defn- choice-input
```
Comment (davrad) [inline src/clj/views/question/create_question_view.clj:30]: Same as above

File: src/clj/views/question/create_question_view.clj [hunk 7]
```diff
@@ -10,196 +10,200 @@
...
+                                  :onclick (str "expert_parakeet.question.create_question_view.delete_element('" div-id "')")}
+                         "-"]]))
+                   possible-solutions)]
+     [:div.input-group
+      (hform/text-field {:class "form-control" :form new-choice-form-id :required true :placeholder "Create new choice"} new-choice-input-id)
+      (hform/submit-button {:class "btn btn-outline-secondary" :form new-choice-form-id} "+")]]))
 
 
 (defn- single-choice-inputs
   [errors question-data]
```
Comment (davrad) [inline src/clj/views/question/create_question_view.clj:72]: Comment is missing

File: src/clj/views/question/create_question_view.clj [hunk 8]
```diff
@@ -10,196 +10,200 @@
...
-                             (optional-error-display :question/possible-solutions errors)
-                             (optional-error-display :question/correct-solutions errors)
-                             (= :question.type/single-choice (question-data :question/type))
-                             (question-data :question/possible-solutions)
-                             (question-data :question/correct-solutions))]))
+  (choice-input errors question-data "single-choice" true))
 
 
 (defn- multiple-choice-inputs
   [errors question-data]
```
Comment (davrad) [inline src/clj/views/question/create_question_view.clj:77]: Comment is missing

File: src/clj/views/question/create_question_view.clj [hunk 9]
```diff
@@ -10,196 +10,200 @@
...
          (h/raw (anti-forgery-field))
-         (hform/submit-button {:class "btn btn-primary"} "submit"))
-       (script "
-expert_parakeet.question.create_question_view.register_question_type_switch('type', " question-types-js-arr ");
-")])))
+         (hform/submit-button {:class "btn btn-primary"} "Submit"))])))
 
 
 (defn- possible-solutions-view
   [{:question/keys [possible-solutions]}]
```
Comment (davrad) [inline src/clj/views/question/create_question_view.clj:184]: Same as above

File: src/cljs/expert_parakeet/question/create_question_view.cljs [hunk 10]
```diff
@@ -3,145 +3,99 @@
     [goog.dom :as dom]))
 
 
+(defn- gen-dom-id
```
Comment (davrad) [inline src/cljs/expert_parakeet/question/create_question_view.cljs:6]: comment is missing

**Review by davrad — CHANGES_REQUESTED** [review]
Looks good to me. Please add comments

**Review by MrDinkleberg — APPROVED** [review]
Das ist schön!

**Review by davrad — APPROVED** [review]

**Pull request merged**